jwst/resample/resample_utils.py

```diff
@@ -144,25 +144,23 @@ def reproject(wcs1, wcs2):
         positions in ``wcs1`` and returns x, y positions in ``wcs2``.
     """
 
-    if isinstance(wcs1, fitswcs.WCS):
-        forward_transform = wcs1.all_pix2world
-    elif isinstance(wcs1, gwcs.WCS):
-        forward_transform = wcs1.forward_transform
-    elif issubclass(wcs1, Model):
-        forward_transform = wcs1
-    else:
-        raise TypeError("Expected input to be astropy.wcs.WCS or gwcs.WCS "
-                        "object or astropy.modeling.Model subclass")
-
-    if isinstance(wcs2, fitswcs.WCS):
-        backward_transform = wcs2.all_world2pix
-    elif isinstance(wcs2, gwcs.WCS):
-        backward_transform = wcs2.backward_transform
-    elif issubclass(wcs2, Model):
-        backward_transform = wcs2.inverse
-    else:
-        raise TypeError("Expected input to be astropy.wcs.WCS or gwcs.WCS "
-                        "object or astropy.modeling.Model subclass")
+    try:
+        forward_transform = wcs1.pixel_to_world_values
+    except AttributeError:
+        if isinstance(wcs1, Model):
+            forward_transform = wcs1
+        else:
+            raise TypeError("Expected input to be of type astropy.wcs.WCS, gwcs.WCS "
+                            f"or astropy.modeling.Model.  Instead it is {type(wcs1)}")
+
+    try:
+        backward_transform = wcs2.world_to_pixel_values
+    except AttributeError:
+        if isinstance(wcs2, Model):
+            backward_transform = wcs2.inverse
+        else:
+            raise TypeError("Expected input to be of type astropy.wcs.WCS, gwcs.WCS "
+                            f"or astropy.modeling.Model.  Instead it is {type(wcs2)}")
```

Comment:
```suggestion
    try:
        forward_transform = wcs1.pixel_to_world_values
    except AttributeError as err:
        from err raise TypeError(
            "Expected input to be of type astropy.wcs.WCS, gwcs.WCS "
            f"or astropy.modeling.Model.  Instead it is {type(wcs1)}"
        )

    try:
        backward_transform = wcs2.world_to_pixel_values
    except AttributeError as err:
        from err raise TypeError(
            "Expected input to be of type astropy.wcs.WCS, gwcs.WCS "
            f"or astropy.modeling.Model.  Instead it is {type(wcs2)}"
        )
```

Reply:
`from err raise TypeError` raised a SyntaxError so I re-wrote and committed another version which checks if ``wcs1`` and ``wcs2`` have the corresponding attributes.

Reply:
I think `reproject()` should still raise a `TypeError`, even though it handles internal `AttributeError`.  This guarantees a consistent behavior, and also is what one would expect from the function if one passes it an object of the wrong type.

Also, I would say it is better to [EAFP](https://docs.python.org/dev/glossary.html#term-EAFP) than to [LBYL](https://docs.python.org/dev/glossary.html#term-LBYL) - more Pythonic.

My bad on the syntax error.  It should be something like

```python
    try:
        forward_transform = wcs1.pixel_to_world_values
        backward_transform = wcs2.world_to_pixel_values
    except AttributeError as err:
        raise TypeError("Input should be a WCS") from err
```

To raise an error from another.  I should have done another PR instead of a code suggestion.  =)

Reply:
You are right. `TypeError` should be raised instead of `AttributeError`. I have made corresponding changes.
Thank you for the explanation and EAFP. I did learn a lot from you in this PR.